### ISSUE TITLE
Form builder - ability to archive (trash/hide) form submissions

### ISSUE THREAD

**Issue by lb-** [wagtail/wagtail#8180]
### Is your proposal related to a problem?

* The form builder feature is an incredible part of Wagtail that allows for easy to build (by users) forms.
* When a form submission is received, often this gets processed (e.g. contact form) by a team or 'actioned', at this point the submission can be deleted but this means it is gone forever.
* The problem is that it would be good to have an 'archive' (or similar) where form submissions can be hidden by default but then still accessed if really needed for historic records.

### Describe the solution you'd like

* AbstractFormSubmission would have a new date field `archived_on`
* AbstractFormSubmission would have a new user field `archived_by`
* By default, new form submissions would have set the `archived_on` `archived_by` fields to `None`
* By default, the form submission listing would not show archived, nor would archived form submissions be exported
* The form submission listing would have a button 'show archived' at the top, very similar to the 'show disabled' in the workflows listing, when clicked the view will change (new URL param) and archived submissions would show, plus there 
* The form submission listing would have a button 'archive' that shows when a form submission is selected
* AbstractFormPage should have a class method / attribute of `disable_ability_to_archive` or similar (default being `False`), if this is `True` the functionality of showing archived submission (including ability to archive) should be hidden from the user. It is possible the name / default could be reversed so that only opt-in is used for this new feature. This also avoids another 'setting' added for this.

<img width="1534" alt="Screen Shot 2022-03-21 at 8 18 42 am" src="https://user-images.githubusercontent.com/1396140/159188452-ca218130-c8f2-4678-aae2-45256f99afb9.png">

Reference - workflows with show disabled button

<img width="1532" alt="Screen Shot 2022-03-21 at 8 18 55 am" src="https://user-images.githubusercontent.com/1396140/159188540-56e0ed4a-06be-48bc-ac32-5e676d7c56f6.png">

### Describe alternatives you've considered

* This can be built with a customisation without too many issues, migrating lots of existing form pages to this would be a bit complex as changed models may be needed though.

### Additional context

These three form builder solutions do NOT provide this functionality, which may indicate that it is not useful

* JotForms allows deletion but not 'archive' https://www.jotform.com/answers/1717595-how-can-i-archive-old-form-responses
* Google allows deletion but not 'archive' https://support.google.com/docs/answer/2917686?hl=en#zippy=%2Cdelete-individual-responses
* Typeform does not allow undelete https://www.typeform.com/help/a/what-happens-to-deleted-results-typeforms-and-accounts-360040196472/

**Comment by vibhuti019** [comment]
I am working on this one

**Comment by vibhuti019** [comment]
I am working on this one

**Comment by lb-** [comment]
See related enhancement to mark form as spam (or more generically 'flagged'). 

https://github.com/wagtail/wagtail/issues/6019

**Comment by vibhuti019** [comment]
Can we implement this using tags ?

**Comment by lb-** [comment]
Not really, tags are more general purpose and are mainly mean to be added in the admin UI.

I also think that flagging and archiving are different things.

Archiving - means it does not show in any default exported lists, index pages etc and must be explicitly requested to show
Flagging - a generic means to highlight a form response that has been received, either automatically (maybe via recaptha) or manually in the UI. Flagged could be used as a way to 'star' something (needs action) or to indicate an issue.

These are related but they are separate things.

Tagging however - would be another thing all together and we would have to be careful we do not use the same Tags table as the ones that could be used in other pages in the UI.

Maybe long term - putting form responses into collections may be useful, but let's bite off one thing at a time.

**Comment by lb-** [comment]
I would think so. 

But best to focus on one thing at a time. Either the flagged one or the archive one. It will be too many changes to do both in the same PR. 

However - it's good to consider the long term somewhat (but not too much haha).

**Comment by lb-** [comment]
It may even be worthwhile doing a first round PR that has the underlying archive functionality of restrictions to the default listing, adds the fields to abstract form page, adds a new method to archive / un-archive. 

Unit tests and functionality for that will be heaps and will be useful. 

Then add the UI buttons as a seperate PR
